llm-base/src/loader.rs

```diff
@@ -0,0 +1,496 @@
+}
+
+/// Used by models to fetch tensors from a loader.
+pub trait TensorLoader<E: std::error::Error> {
+    /// Loads a tensor from the loader.
+    fn load(&mut self, name: &str, ne: &[usize]) -> Result<ggml::Tensor, E>;
+    /// Finish loading the model, and extract all of the state from the loader.
+    fn finish(self) -> (ggml::Context, HashMap<String, ggml::Tensor>, Option<Mmap>);
+}
+
+/// Load an arbitrary GGML model.
+pub fn load<M: Model>(
+    path: impl AsRef<Path>,
+    prefer_mmap: bool,
+    n_context_tokens: usize,
+    mut load_progress_callback: impl FnMut(LoadProgress),
+) -> Result<M, LoadError> {
+    let main_path = path.as_ref();
+
+    let paths = util::find_all_model_files(main_path)?;
+    if paths.len() != 1 {
+        return Err(LoadError::MultipartNotSupported { paths });
+    }
+
+    let file = File::open(main_path).map_err(|e| LoadError::OpenFileFailed {
+        source: e,
+        path: main_path.to_owned(),
+    })?;
+    let mut reader = BufReader::new(&file);
+
+    let path = path.as_ref().to_owned();
+
+    (load_progress_callback)(LoadProgress::PartLoading {
+        file: &path,
+        current_part: 0,
+        total_parts: 1,
+    });
+
+    let mut loader = Loader::new(load_progress_callback);
+
+    ggml_format::load_model(&mut reader, &mut loader)
+        .map_err(|err| LoadError::from_format_error(err, path.clone()))?;
+
+    let Loader {
+        hyperparameters,
+        vocabulary,
+        tensors,
+        mut load_progress_callback,
+        container_type,
+        ..
+    } = loader;
+
+    let use_mmap = prefer_mmap && container_type.support_mmap();
+
+    let ctx_size = tensors
+        .values()
+        .map(|ti| {
+            ggml::Tensor::C_TYPE_SIZE
+                + ggml::OBJECT_SIZE
+                + if use_mmap { 0 } else { ti.calc_size() }
+        })
+        .sum::<usize>();
+    (load_progress_callback)(LoadProgress::ContextSize { bytes: ctx_size });
+    let context = ggml::Context::init(ctx_size, !use_mmap);
+
+    let mmap = if use_mmap {
+        let file = File::open(&path)?;
+        Some(unsafe { Mmap::map(&file)? })
+    } else {
+        None
+    };
+
+    struct MmapCompatibleLoader<'a> {
+        path: PathBuf,
+        file: File,
+        tensors: HashMap<String, TensorInfo>,
+        context: ggml::Context,
+        mmap: Option<Mmap>,
+        load_progress_callback: &'a mut dyn FnMut(LoadProgress),
+        loaded_tensors: HashMap<String, ggml::Tensor>,
+    }
+    impl TensorLoader<LoadError> for MmapCompatibleLoader<'_> {
+        fn load(&mut self, name: &str, ne: &[usize]) -> Result<ggml::Tensor, LoadError> {
+            let info = self
+                .tensors
+                .get(name)
+                .ok_or_else(|| LoadError::UnknownTensor {
+                    path: self.path.clone(),
+                    tensor_name: name.to_owned(),
+                })?;
+
+            let ctx = &self.context;
+            let mut tensor = match ne.len() {
```

Comment:
Should this be checked?

```suggestion

            let dims = ne.len();
            if dims != info.n_dims {
                return Err(LoadError::InvariantBroken {
                    path: self.path.clone(),
                    invariant: format!(
                        "the tensor {name} should have {} dimensions, not {dims}",
                        info.n_dims
                    ),
                });
            }

            let ctx = &self.context;
            let mut tensor = match dims {
```

Reply:
That seems reasonable to me, assuming that does what it looks like (you can never be sure with those short variable names...)

If it works, commit it!